PyInstaller/archive/readers.py

```diff
@@ -143,15 +143,23 @@ def checkmagic(self):
         else:
             self.lib.seek(0, 2)
         filelen = self.lib.tell()
-        if self.length:
-            self.lib.seek(self.start + self.length - self._cookie_size, 0)
-        else:
-            self.lib.seek(-self._cookie_size, 2)
-        (magic, totallen, tocpos, toclen, pyvers, pylib_name) = struct.unpack(
+        found = False
+        for i in range(4096):
+            self.lib.seek(filelen-self._cookie_size, 0)
+            (magic, totallen, tocpos, toclen, pyvers, pylib_name) = struct.unpack(
```

Comment:
Why not simply read 4096+cookie_size bytes and use `.rfind()` to find the magic, like this:
```
  seek(max(0, filelen-4096-self._cookie_size)) # file may be smaller
  buf = read(min(filelen, 4096.self._cookie_size) # file may be shorter
  pos = buf.rfind(MAGIC)
  if pos < 0:
   raise RunoitemError(...)
  magic, toallen, ... = unpack(_copkie_format, buf)
```